### PR TITLE
New extension: HenTalk

### LINKED ISSUE #18550: Reopen F**ku cc ticket
### Source name

F**ku.cc

### Source link

https://fakku.cc

### Source language

English

### Other details

There was a source request for this in 2022 but the page got shut down and now is back up; the reply for that request was to use lanraragi. I don't really know how to set this up (and would rlly appreciate it if someone would link a detailed guide for this), but my guess is that it would be easier and a lot more practical for most of users to have a direct extension.

Thank you for reading

### Acknowledgements

- [X] I have checked that the extension does not already exist on the [website extensions list](https://tachiyomi.org/extensions/) or the app.
- [X] I have checked that the extension is not on [the removed sources list](https://github.com/tachiyomiorg/tachiyomi-extensions/blob/master/REMOVED_SOURCES.md).
- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [x] I have checked that the extension does not already exist by searching the [GitHub repository](https://github.com/tachiyomiorg/tachiyomi-extensions/) and verified it does not appear in the code base.
- [X] I have written a meaningful title with the source name.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by BrutuZ**
Closes #18550

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio

### CODE REVIEW

**Comment by alessandrojean** [comment]
Haven't reviewed it properly yet, but can you remove the log statements from the source class? The ones in the deep linking activity are fine to be kept.

**Comment by BrutuZ** [comment]
Totally overlooked that 😅. I even had the Search and Replace already filled with the RegEx but forgot to "do the thing" after the device tests

File: src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+
+    private fun buildAdvQuery(
+        url: HttpUrl.Builder,
+        query: String,
+        filters: FilterList,
+    ) {
+        var searchString = query
+        filters.filter { it.state !in listOf(null, "", 0) }.forEach { filter ->
+            with(filter.name) {
+                when {
```
Comment (alessandrojean) [inline src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt:102]: Kotlin has proper support to `==`. This whole `when` can be simplified like:

```kotlin
when (filter.name) {
  "Title", "Artist", "Circle" -> { /* ... */ }
  "Tag" -> { /* ... */ }
  else -> { /* ... */ }
}
```

File: src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        }
+        if (searchString.trim().isNotEmpty()) url.setQueryParameter("q", searchString.trim())
+    }
+
+    override fun mangaDetailsParse(document: Document) = SManga.create().apply {
+        val art =
+            document.select(".metadata a[href^='/artists/'], article > a div.artists")
+                .map {
+                    it.text()
+                }.distinct().joinToString()
```
Comment (alessandrojean) [inline src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt:150]:
```suggestion
            document.select(".metadata a[href^='/artists/'], article > a div.artists")
                .map { it.text() }.distinct().joinToString()
```

File: src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    override fun chapterListParse(response: Response): List<SChapter> {
+        val document = response.asJsoup()
+
+        return listOf(
+            SChapter.create().apply {
+                setUrlWithoutDomain(response.request.url.encodedPath)
+                val title: String = document.selectFirst(".metadata h1.title")!!.text() ?: ""
+                val ch = Regex("(?<!20\\d\\d-)\\b[\\d.]{1,4}$").find(title)?.value?.trim('.') ?: "1"
+                name = "$ch. $title"
+                chapter_number = ch.toFloat()
```
Comment (alessandrojean) [inline src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt:221]: Doesn't need to set it, let the app parse it.

File: src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+
+    override fun chapterListSelector() = popularMangaSelector()
+
+    override fun pageListRequest(chapter: SChapter) = GET("$baseUrl${chapter.url}", headers)
+
+    override fun pageListParse(document: Document): List<Page> {
+        val totalPages = document.selectFirst(".metadata .pages > td:nth-child(2)")!!.text()
+            .toInt()
+        if (totalPages == 0) {
+            throw UnsupportedOperationException("Error: Empty pages (try Webview)")
```
Comment (alessandrojean) [inline src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt:238]: Just return an `emptyList()` in this case, as the app already handles this. The user can open in WebView if needed, there isn't need for a custom error.

File: src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+                document.selectFirst("figure.thumbnail")!!.attr("src").replace("/896/", "/320/")
+            val cdn = cover.substringBeforeLast('/').substringBeforeLast('/')
+            for (i in 1..totalPages) {
+                srcs.add("$cdn/$i.jpg")
+            }
+        }
+
+        return srcs.mapIndexed { idx, img ->
+            Page(idx, imageUrl = img.replace("/320.webp", ".jpg"))
+        }
```
Comment (alessandrojean) [inline src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt:254]:
```suggestion
        val previewImages = document.select(".previews > .preview img")
            .map { if (it.hasAttr("data-src")) it.absUrl("data-src") else it.absUrl("src") }
            
        if (previewImages.isEmpty()) {
            val cover = document.selectFirst("figure.thumbnail")!!
                .attr("src").replace("/896/", "/320/")
            val cdnUrl = cover.substringBeforeLast("/")
                .substringBeforeLast("/")
            
            return List(totalPages) { Page(it, imageUrl = "$cdn/${i + 1}.jpg" }
        }
        
        return previewImages.mapIndexed { index, imageSrc -> {
            Page(index, imageUrl = imageSrc.replace("/320.webp", ".jpg"))
        }
```

File: src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalkUrlActivity.kt [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+import android.app.Activity
+import android.content.ActivityNotFoundException
+import android.content.Intent
+import android.os.Bundle
+import android.util.Log
+import kotlin.system.exitProcess
+
+/**
+ * Springboard that accepts https://fakku.cc/archive/xxxxx intents and redirects them to
```
Comment (alessandrojean) [inline src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalkUrlActivity.kt:11]:
```suggestion
 * Activity that accepts https://fakku.cc/archive/xxxxx intents and redirects them to
```

File: src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalkWebViewInterceptor.kt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+import java.io.IOException
+import java.util.concurrent.CountDownLatch
+
+class HenTalkWebViewInterceptor : Interceptor {
+
+    override fun intercept(chain: Interceptor.Chain): Response {
+        val request = chain.request()
+        val response = chain.proceed(request)
+
+        if (response.headers("Content-Type").any { it.contains("text/html") }) {
```
Comment (alessandrojean) [inline src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalkWebViewInterceptor.kt:22]: There's only one `Content-Type` header. If the site is sending multiple, they have an issue.

```suggestion
        if (response.headers["Content-Type"]!!.contains("text/html")) {
```

**Review by alessandrojean — CHANGES_REQUESTED** [review]

File: src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt (hunk 3 above)
Comment (BrutuZ) [reply src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt:221]: It'll handle all the way from `val ch = ...`?

Comment (alessandrojean) [reply src/en/hentalk/src/eu/kanade/tachiyomi/extension/en/hentalk/HenTalk.kt:221]: Can remove just the line that sets `chapter_number`.

**Comment by BrutuZ** [comment]
Worked for the OG Koushoku layout they had going, not it's pointless since the site just switched to LANraragi

**Pull request closed without merging**